### ISSUE TITLE
sensu-api just quits once in a while (after sensu-server restart usually)

### ISSUE THREAD

**Issue by FlorinAndrei** [sensu/sensu#815]
I've seen this today on version 0.16.0, but it's been happening ever since I've started using Sensu, a few versions back.

Basically, sensu-api just quits. This seems to happen more often if I restart sensu-server, so I acquired the habit of restarting sensu-api too a few seconds after restarting sensu-server - that seems to help somewhat.

These are the messages I've seen in the logs today:

```
{"timestamp":"2014-11-03T11:02:00.103424-0800","level":"info","message":"GET /clients","remote_address":"127.0.0.1","user_agent":null,"request_method":"GET","request_uri":"/clients","request_body":""}
{"timestamp":"2014-11-03T11:02:00.107152-0800","level":"info","message":"GET /events","remote_address":"127.0.0.1","user_agent":null,"request_method":"GET","request_uri":"/events","request_body":""}
{"timestamp":"2014-11-03T11:02:10.108778-0800","level":"info","message":"GET /info","remote_address":"127.0.0.1","user_agent":null,"request_method":"GET","request_uri":"/info","request_body":""}
{"timestamp":"2014-11-03T11:02:10.109866-0800","level":"fatal","message":"transport connection error","error":"rabbitmq channel closed"}
{"timestamp":"2014-11-03T11:02:10.109973-0800","level":"warn","message":"stopping"}
{"timestamp":"2014-11-03T11:02:10.610888-0800","level":"warn","message":"stopping reactor"}
```

I cannot be sure if these same messages were printed during previous crashes, I was too busy to look at the logs until today, sorry.

If I do "service sensu-api restart", it resumes working without fuss.

I'm also using whatever is the latest Uchiwa and RabbitMQ in the repos, all on Ubuntu 14.04 fully updated. A couple dozen clients spread through the cloud in various geographies, usually running same Sensu version as the server.

**Comment by srodrig0209** [comment]
I also just experienced this. I terminated a secondary dev instance and saw the following:

```
=INFO REPORT==== 14-Jan-2015::11:37:48 ===
accepting AMQP connection <0.13516.18>

=WARNING REPORT==== 14-Jan-2015::11:37:48 ===
closing AMQP connection <0.13516.18> :
connection_closed_abruptly

=INFO REPORT==== 14-Jan-2015::11:37:50 ===
Statistics database started.

=INFO REPORT==== 14-Jan-2015::11:37:50 ===
rabbit on node 'rabbit@dev-sensu-instance' down

=ERROR REPORT==== 14-Jan-2015::11:37:51 ===
Channel error on connection <0.12706.14> (vhost: '/sensu', user: 'sensu'), channel 1:
{amqp_error,not_found,"no queue 'keepalives' in vhost '/sensu'",
            'queue.declare'}

=INFO REPORT==== 14-Jan-2015::11:37:52 ===
closing AMQP connection <0.12706.14> 

=INFO REPORT==== 14-Jan-2015::11:37:52 ===
node 'rabbit@dev-sensu-instance' down: connection_closed
```


```
{"timestamp":"2015-01-14T11:37:51.798170-0500","level":"fatal","message":"transport connection error","error":"rabbitmq channel closed"}
{"timestamp":"2015-01-14T11:37:51.798301-0500","level":"warn","message":"stopping"}
{"timestamp":"2015-01-14T11:37:52.299001-0500","level":"warn","message":"stopping reactor"}
```

**Comment by dblessing** [comment]
I just encountered the same thing. In our case, this behavior is triggered by a Redis failover. When Redis master fails the sensu-server unsubscribes from the results and keepalive queues in RabbitMQ, which ultimately leads to the failure of sensu-api.

I think both the RabbitMQ and the Redis transports in Sensu need to treat transport errors as non-fatal, maybe pause for a moment and re-try the connections. In an HA configuration it's perfectly reasonable to expect that the error scenario will eventually correct itself.

I see this in sensu-server:
```
{"timestamp":"2015-01-16T08:23:26.105911-0600","level":"warn","message":"reconnecting to redis"}
{"timestamp":"2015-01-16T08:23:26.106250-0600","level":"warn","message":"unsubscribing from keepalive and result queues"}
{"timestamp":"2015-01-16T08:23:26.106595-0600","level":"debug","message":"not currently master"}
```

followed by the sensu-api logs that @FlorinAndrei posted in the original issue:

```
{"timestamp":"2014-11-03T11:02:10.109866-0800","level":"fatal","message":"transport connection error","error":"rabbitmq channel closed"}
{"timestamp":"2014-11-03T11:02:10.109973-0800","level":"warn","message":"stopping"}
{"timestamp":"2014-11-03T11:02:10.610888-0800","level":"warn","message":"stopping reactor"}
```

and the RabbitMQ logs that @benh57 posted:

```
=ERROR REPORT==== 9-Dec-2014::16:58:32 ===
connection , channel 1 - soft error:
{amqp_error,not_found,"no queue 'keepalives' in vhost 'sensu'",
'queue.declare'}
```

**Comment by dblessing** [comment]
This seems to be somewhat related to #599

**Comment by portertech** [comment]
The API `transport_info()` https://github.com/sensu/sensu/blob/master/lib/sensu/api.rb#L204 needs to check that the queue exists before querying for its stats.

**Comment by sarkis** [comment]
We are also seeing this issue. The sensu-api log shows the following error when we start noticing the issues with uchiwa:

{"timestamp":"2015-03-07T03:35:04.853583+0000","level":"fatal","message":"transport connection error","error":"rabbitmq channel closed"}

I logged into the sensu server and looked at the sensu-api status: sudo service sensu-api status

I got a message that says pid file exists but service is not running. So I started up the sensu-api and seems to have resolved at least for now...

**Comment by jcmcken** [comment]
I was also seeing this issue with the same symptoms: server and API randomly die, RMQ logs show the error about the keepalives queue. In my case, I think this was related to me upgrading and downgrading Sensu to test various things without flushing Redis. Not sure if this is everyone's issue, but you can try shutting down all the Sensu services, running a ``redis-cli FLUSHALL``, and then restarting Sensu. That seemed to fix the issue for me (at least so far, still watching it).

**Comment by portertech** [comment]
@sarkis @jcmcken currently, you can use the following transport configuration to reconnect on error (such as channel closed):

```
{
  "transport": {
    "reconnect_on_error": true
  }
}
~~~

**Comment by portertech** [comment]
We could make this the default (`true`).

**Comment by FlorinAndrei** [comment]
@portertech Please do, using reasonable timing settings.

**Comment by pnixon-kwicr** [comment]
I have two sites that continuously fall offline due to this, which config file does that transport snippet live in?

**Comment by portertech** [comment]
@pnixon-kwicr you could create, for example, `/etc/sensu/conf.d/transport.json`, containing:

```
{
  "transport": {
    "reconnect_on_error": true
  }
}
```

**Comment by sarkis** [comment]
I've confirmed this is fixed now in versions > 0.18.0 - as reconnect_on_error defaults to true. Thanks @portertech

**Comment by countryHick** [comment]
I'm getting this with sensu-client alot lately -- can this apply there as well - i'll check the rabbit logs and see what i can find...

**Comment by mkobel** [comment]
I'm getting this with sensu-api 0.20.0. I could not yet find any log entry for the reason.

```
{"timestamp":"2015-07-19T06:43:19.466125+0200","level":"warn","message":"received signal","signal":"TERM"}
{"timestamp":"2015-07-19T06:43:19.466327+0200","level":"warn","message":"stopping"}
{"timestamp":"2015-07-19T06:43:19.967221+0200","level":"warn","message":"stopping reactor"}
```

**Comment by countryHick** [comment]
i found that puppet was recycling the sensu-client EACH run (due to bug in
version, fixed in master) - and it's alot better now that i'm running
master version...

**Comment by mkobel** [comment]
Did you upgrade sensu or its puppet module?

**Comment by mkobel** [comment]
I assume I just found the reason: #974 The LXC server outside the container stops the sensu-api and the sensu-server with its puppet configuration.

**Comment by calebhailey** [comment]
Closing because it looks like this was resolved w/ #974

**Comment by benh57** [comment]
Nah, we aren't using any sort of openvz container and i've seen this often in the past.  However, i haven't seen this issue in a while. So i'll re-open only if i do.
